### PR TITLE
Search users

### LINKED ISSUE #530: Find contact by typing in username or name
Show a list of members by typing in *username*, *firstName*, *lastName* or *displayName*.

- The API is already implemented at #531
- Markup can be basically the same as at contacts tab on the user profile
- First iteration doesn't need pagination, just hard limit (e.g. 10-30 or something)
- ~First iteration doesn't need to find from the middle of the name, only from the beginning~ We'll not implement partial search at all, see https://github.com/Trustroots/trustroots/pull/531#issuecomment-430018690
- Should not be case sensitive
- In future this page would hold also FB your friends who are also on TR (#392).
- Might need adding another Mongo index to *firstName*, *lastName* and *displayName*.

In navigation - on mobile:
image

In navigation - on desktop:
image

Label could be "Find friends".

The goal of this feature is:
- to find your existing contacts easier from Trustroots so that you can 1) add them as contacts 2) write to them.
- to not to find hosts or members to meet.

### PR DESCRIPTION
**Pull request by Akronix**
#### Proposed Changes

* Add a new /search/users endpoint to look for users by username. 
* Coded UsersList, Avatar and SearchUsers components in React.
* Added a button to switch from map search to user search and vice versa.

First approach to resolve #530

#### Testing Instructions

1. Run the dev server and go to http://localhost:3000/search/users.
2. Write an existing username, like admin1 and press the search button.

#### Current status
Screenshot of current status:
image

#### Still needs to be done

- [x] Show profile pics
- [ ] Show other info like living location, language, gender, last seen,.. Depends on #944
- [x] Pressing Enter key, do submit search.
- [ ] Fuzzy search. Depends on #944
- [x] Check they are well translated and remove angular left-overs.
- [ ] add tooltips (?)
- [ ] Use `search?=` http parameter to init defaultProps.
- [ ] Add a link from the /profile/<username>/contacts page to /search/users
- [x] Final: remove console.logs and prepopulated `admin1`

### CODE REVIEW

File: modules/search/client/components/UsersList.js [hunk 1]
```diff
@@ -1,16 +1,16 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 
-const Avatar = ({ avatarSize, avatarSource, userData }) => (
+const Avatar = ({ avatarSize, avatarSource, displayName, profileUrl }) => (
```
Comment (simison) [inline modules/search/client/components/UsersList.js]: Let's start this as an external component so that it becomes reusable (we'll literally need it everywhere from now on).

Our previous client-side Avatar directive is terrible; we should clearly move all that logic to the backend:

https://github.com/Trustroots/trustroots/blob/bbacc11d05923134a8fc140472473b43ded17716/modules/users/client/directives/tr-avatar.client.directive.js

**Comment by guaka** [comment]
When I go to http://localhost:3000/search/users the page keeps reloading every second or so.

**Comment by petr-hajek** [comment]
I think to be releasable we need to do just these:
- [ ] Show profile pics + remove non-implemented info
- [ ] Pressing Enter key, do submit search.
- [ ] Final: remove console.logs and prepulated `admin1`
Rest can be done in later iterations.

I am mostly thinking if the button to get to the search is intuitive. I think I wouldn't notice as regular user. Maybe we could add a text link somewhere else also?

**Comment by Akronix** [comment]
> I think to be releasable we need to do just these:
> 
>     * [ ]  Show profile pics + remove non-implemented info
> 
>     * [ ]  Pressing Enter key, do submit search.
> 
>     * [ ]  Final: remove console.logs and prepulated `admin1`
>       Rest can be done in later iterations.
> 

Agree!

> 
> I am mostly thinking if the button to get to the search is intuitive. I think I wouldn't notice as regular user. Maybe we could add a text link somewhere else also?

Yes, I forgot to say! There could be a link from the contacts page.

**Comment by simison** [comment]
>* [ ]  Show profile pics + remove non-implemented info

Profile images might require a bit more work to work so I'd be okay to hide it for the first iteration.

**Comment by Akronix** [comment]
> > * [ ]   Show profile pics + remove non-implemented info
> 
> Profile images might require a bit more work to work so I'd be okay to hide it for the first iteration.

Really? But I think it is the most important info to display for a search users feature.

I might have some time today to do this:

>  [ ]  Pressing Enter key, do submit search.

**Comment by simison** [comment]
> >   Profile images might require a bit more work to work so I'd be okay to hide it for the first iteration.

>Really? But I think it is the most important info to display for a search users feature.

I think so too and we definitely need to add it eventually. I meant that since it's quite some work either at client side or at the backend to make all the possible avatar sources work (none/uploaded/gravatar/facebook), it would make sense to drop it from the first version just to get something already functional published.

Creating an avatar component in React is pretty high on priorities though, since it's needed literally everywhere.

File: modules/search/client/components/UsersList.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+Avatar.propTypes = {
+  avatarSize: PropTypes.number,
+  avatarSource: PropTypes.string,
+  displayName: PropTypes.string,
+  profileUrl: PropTypes.string
+};
+
+const User = ({ userData }) => {
```
Comment (simison) [inline modules/search/client/components/UsersList.js]: Would be nice to extract this to an external reusable component (we could reuse it e.g. for the contacts list). That's okay for another PR too.

File: modules/search/client/components/UsersList.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+User.propTypes = {
+  userData: PropTypes.object
+};
+
+
+export default function UsersList({ users }) {
+  const usersEvenComponent = [];
+  const usersUnEvenComponent = [];
```
Comment (simison) [inline modules/search/client/components/UsersList.js]: We can pretty easily achieve this layout by using CSS flex or CSS grid instead of using Bootstrap columns — but that can be another PR as well.

File: modules/search/client/components/search.api.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export async function searchUsers(query) {
+  const response = await fetch(`/api/users?search=${query}`, {
```
Comment (simison) [inline modules/search/client/components/search.api.js]: We can now rebase this PR and switch to use `axios` instead of `fetch`, see docs:

https://github.com/Trustroots/trustroots/blob/master/docs/React.md#api-calls-from-react

Comment (Akronix) [reply modules/search/client/components/search.api.js]: Is this service already anywhere or is it going to be anywhere anytime soon?

**Comment by Akronix** [comment]
> > > Profile images might require a bit more work to work so I'd be okay to hide it for the first iteration.
> 
> > Really? But I think it is the most important info to display for a search users feature.
> 
> I think so too and we definitely need to add it eventually. I meant that since it's quite some work either at client side or at the backend to make all the possible avatar sources work (none/uploaded/gravatar/facebook), it would make sense to drop it from the first version just to get something already functional published. Lack of avatar isn't blocking anyone from using this feature.
> 
> Creating an avatar component in React is pretty high on priorities though, since it's needed literally everywhere.

OK, then. I can remove all the avatar stuff, but somebody should be responsible to take that task and start coding it. 

Maybe that person wants to start from my code?

**Comment by simison** [comment]
> Maybe that person wants to start from my code?

Yep I think this section is great to get it started but separate PR would be better; this one is quite big already.

**Comment by simison** [comment]
Opened an issue for avatar: https://github.com/Trustroots/trustroots/issues/982

File: modules/search/client/components/search.api.js (hunk 4 above)
Comment (simison) [reply modules/search/client/components/search.api.js]: I don't think was added yet. Axios library is in already tho.

Comment (simison) [reply modules/search/client/components/search.api.js]: Adding this in https://github.com/Trustroots/trustroots/pull/984

File: modules/search/client/components/UsersList.js (hunk 1 above)
Comment (mrkvon) [reply modules/search/client/components/UsersList.js]: See #997

Comment (simison) [reply modules/search/client/components/UsersList.js]: https://github.com/Trustroots/trustroots/pull/997 is now merged and this can be changed to use new Avatar component instead. 💪

File: modules/search/client/components/SearchUsers.component.js [hunk 5]
```diff
@@ -91,8 +92,10 @@ class SearchUsers extends React.Component {
     this.setState({ resolved: false });
     console.log('Searching for: ' + this.state.searchQuery);
 
+    // eslint-disable-next-line react/no-access-state-in-setstate
     const userResults = await searchUsers(this.state.searchQuery);
```
Comment (Akronix) [inline modules/search/client/components/SearchUsers.component.js]: I'm getting this eslint error in that line:
`95:43  error  Use callback in setState when referencing the previous state  react/no-access-state-in-setstate`
I have no idea on how to fix it. Any idea @simison @mrkvon ?
This is the eslint docs about it: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-access-state-in-setstate.md

Comment (mrkvon) [reply modules/search/client/components/SearchUsers.component.js]: This is quite peculiar.
This rule takes care specifically of:
```jsx
// incorrect, can lead to state inconsistency
this.setState({ isFoo: !this.state.isFoo }); // flipping a random boolean

// correct
this.setState((previousState) => ({ isFoo: !previousState.isFoo }));
```
@Akronix Your link is describing the reasons for this.

Perhaps try
```
this.setState(() => ({ ... }))
```
but it shouldn't be necessary.

In a while, if not resolved, I can try locally.

File: modules/search/client/components/SearchUsers.component.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      'borderLeftStyle': 'solid',
+      'borderColor': '#eee' };
+
+    const switchToSearchPlaces = (
+      <span className="input-group-btn"
+        style={borderStyle}>
+        <a type="button"
+          href="/search"
+          className="btn btn-lg btn-default"
+          tooltip-enable="::search.screenWidth >= 768"
```
Comment (simison) [inline modules/search/client/components/SearchUsers.component.js]: A little bit of Angular tooltip directive here that can be removed.

File: modules/search/client/components/UsersList.js (hunk 2 above)
Comment (simison) [reply modules/search/client/components/UsersList.js]: Done in https://github.com/Trustroots/trustroots/pull/1009

File: modules/search/client/components/SearchUsers.component.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    this.setState({ resolved: true, users: userResults.data });
+    console.log(this.state.users);
+  };
+
+  clickClear() {
+    this.setState({ searchQuery: '' });
+  }
+
+  render() {
+    console.log('SearchUsers: render');
```
Comment (simison) [inline modules/search/client/components/SearchUsers.component.js]: Let's remember to remove `console.log`s from the PR before merging.

File: modules/search/client/components/SearchUsers.component.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+    const borderStyle = { 'borderLeftWidth': '1px',
+      'borderLeftStyle': 'solid',
+      'borderColor': '#eee' };
```
Comment (simison) [inline modules/search/client/components/SearchUsers.component.js]: Let's remember to move this style object to Less file with other our styles.

File: modules/search/client/components/SearchUsers.component.js (hunk 7 above)
Comment (Akronix) [reply modules/search/client/components/SearchUsers.component.js]: sure :+1:

Comment (mrkvon) [reply modules/search/client/components/SearchUsers.component.js]: [eslint's `no-console`](https://eslint.org/docs/rules/no-console) can be interesting for us.

Just disable it explicitly whenever we really need it with `eslint-disable-line`.

File: modules/search/client/components/SearchUsers.component.js (hunk 6 above)
Comment (Akronix) [reply modules/search/client/components/SearchUsers.component.js]: Well, I left all the angular stuff just to make sure they were being correctly translated to react (at least, the basic ones), but I can remove them very easily.

File: modules/search/client/components/SearchUsers.component.js (hunk 8 above)
Comment (Akronix) [reply modules/search/client/components/SearchUsers.component.js]: hum, something to add to the todo list.

File: modules/search/client/components/SearchUsers.component.js (hunk 5 above)
Comment (mrkvon) [reply modules/search/client/components/SearchUsers.component.js]: Still there? For me `npm run lint` passes.

Comment (mrkvon) [reply modules/search/client/components/SearchUsers.component.js]: Two lines below, use `this.setState(() => ({ resolved: true, users: userResults.data }))` instead. `const userResults = ...` references `this.state` and that's what this rule is worried about.

I haven't wrapped my head around this particular example, but at least the lint error disappears! :slightly_smiling_face:

Comment (simison) [reply modules/search/client/components/SearchUsers.component.js]: I think this section should be structured like this for it to work as expected and make the linter happy (untested):

```js
this.setState({ resolved: false }, async () => {
  searchUsers(this.state.searchQuery)
    .then((data) => {
      // Do something in success
    })
    .catch((error) => {
      // Do something in errors
    })
    .finally(() => {
      this.setState({
        resolved: true,
      });
    });
});
```

Comment (simison) [reply modules/search/client/components/SearchUsers.component.js]: That said, perhaps kinda `resolved` (I can regognice Angularism) could be renamed to `isQuerying`, `isSearching` or similar?

File: modules/search/client/components/SearchUsers.component.js (hunk 7 above)
Comment (simison) [reply modules/search/client/components/SearchUsers.component.js]: Valid point @mrkvon ! 👍 :shipit:

Comment (simison) [reply modules/search/client/components/SearchUsers.component.js]: @mrkvon added the rule in https://github.com/Trustroots/trustroots/pull/1015

File: modules/search/client/components/SearchUsers.component.js (hunk 6 above)
Comment (Akronix) [reply modules/search/client/components/SearchUsers.component.js]: @simison I'm thinking to better leave the tooltip directives just in case we got something similar working in the future for React? At the end, they don't really annoy so much, do they?

**Comment by Akronix** [comment]
We got the basic points done:

- [x] Show profile pics
- [x] Pressing Enter key, do submit search.
- [x] Check they are well translated and remove angular left-overs.
- [x] Final: remove console.logs and prepopulated `admin1`

This is ready to review from my side :tada:

File: modules/search/client/components/SearchUsers.component.js (hunk 6 above)
Comment (simison) [reply modules/search/client/components/SearchUsers.component.js]: It's really better to remove anything that's not used. 🙂 It can be confusing to have dead code around. If we create tooltips in future (probably not as they're not great user interface) it'll look different to Angular directives so they won't work "out of the box" with this syntax anyway.

File: modules/search/client/components/SearchUsers.component.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    <div>
+      {!isSearching?
+        <div className="content-wait"
+          role="alertdialog"
+          aria-busy="true"
+          aria-live="assertive">
+          <small>Wait a moment...</small>
+        </div>
+        :
+        (users && users.length > 0? userList : noUsers)
```
Comment (simison) [inline modules/search/client/components/SearchUsers.component.js]: I really like this variable -way of rendering components. 👍

File: modules/search/client/components/SearchUsers.component.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+          </h4>
+        </div>
+      </div>
+    </div>
+  );
+
+  const noUsers = (
+    <div className="row content-empty">
+      <i className="icon-3x icon-users"></i>
+      <h4>No users found with this name.</h4>
```
[the remark this answers is not in the document]
Comment (Akronix) [reply modules/search/client/components/SearchUsers.component.js]: Done!

File: modules/search/client/views/search-input.client.view.html [hunk 11]
```diff
@@ -1,6 +1,18 @@
...
+      <a type="button"
+        href="/search/users"
+        class="btn btn-lg btn-default"
+        tooltip-enable="::search.screenWidth >= 768"
+        tooltip-placement="bottom"
+        uib-tooltip="Go to search users"
+        aria-label="Go to search users by username">
+        <i class="icon-users"></i>
+      </a>
+    </span>
```
Comment (simison) [inline modules/search/client/views/search-input.client.view.html]: I feel like we could take a little different route with discoverability/linking map search and user search to each other.

While these function well I think we can improve them to work around these drawbacks:

- It's not clear what happens from the button or that it even _is_ a button:

screenshot 2018-12-30 at 20 41 14
screenshot 2018-12-30 at 20 41 24

- It looks like it's part of the search input and there to provide visual symbol for thes earch rather than it's a way to access completely different major interface.

The tooltip helps little but we cannot rely on that; especially on mobile where we don't have hover-state. Seems like it has issues also on the desktop:

screenshot 2018-12-30 at 20 41 32

We can assume most folks not seeing these tooltips so we need to build UI that's clear without them and hide only additional, optional information in tooltips.

The intent at the user search side is little bit more clearer but I'd say it has similar issues and it makes the search input rather busy with all the icons:

image

I would suggest to hide these UIs for now and e.g. add link to this feature only behind user-dropdown. That's not very discoverable of course so we'll need to work more on it, but we can have different PR for that so that we don't slow down merging this.

I know you're AFK right now; hope you don't mind if I go ahead and do those PRs and we can still add these inputs in some format if we so conclude! :-)

**Comment by simison** [comment]
Aight, so I:
- removed icon-cross links from map search to here; I'll add them back in a separate PR in slightly different format
- Restructured just a little bit `isLoading` state
- Removed "admin1" demo query

File: modules/search/client/views/search-input.client.view.html (hunk 11 above)
Comment (simison) [reply modules/search/client/views/search-input.client.view.html]: Work in progress here: https://github.com/Trustroots/trustroots/pull/1020

File: modules/search/client/config/search.client.routes.js [hunk 12]
```diff
@@ -80,6 +80,15 @@
         data: {
           pageTitle: 'Search'
         }
+      }).
+      state('search-users', {
```
Comment (Akronix) [inline modules/search/client/config/search.client.routes.js]: @simison  I find this search-users /search/members inconsistent. Shall we move everything to members naming convention?
